File: lazypx4/satellite.py
"""Satellite-image snapshot for the position map ([n] -> [i]).

Downloads a Web-Mercator tile from Esri "World Imagery" (no API key) covering
the vehicle and its home point (and, if one is loaded, the [o] KML overlay's
fence/waypoints too), annotates it with pins, the fence boundary, waypoint
markers and a scale bar if Pillow is installed, and writes the image plus a
JSON sidecar to ``settings.map_dir``. Runs on a background thread; needs
internet access.
"""

from __future__ import annotations

import json
import math
import os
import threading
import time
from datetime import datetime, timezone

from .config import (
    MAP_DOWNLOAD_TIMEOUT,
    MAP_IMAGE_MIN_SPAN_M,
    MAP_IMAGE_PAD,
    MAP_IMAGE_SIZE,
    MAP_RANGE_MAX_M,
    settings,
)
from .eventlog import log_command, log_error, log_info, log_warn
from .state import state
from .util import clamp


def _web_mercator(lat, lon):
    radius = 6378137.0
    x = radius * math.radians(lon)
    y = radius * math.log(
        math.tan(math.pi / 4.0 + math.radians(clamp(lat, -85.0, 85.0)) / 2.0)
    )
    return x, y


def start_satellite_download():
    with state.lock:
        if state.map_download_active:
            log_warn("A satellite-image download is already running")
            return False

        if not state.global_pos_valid:
            state.map_download_status = "ERROR"
            state.map_download_error = (
                "no GPS position (need a GPS fix / GLOBAL_POSITION_INT)"
            )
            state.map_download_path = ""
            log_error("Satellite image: no GPS position available")
            return False

        robot_lat = state.global_lat
        robot_lon = state.global_lon

        if state.local_origin_set:
            home_lat = state.local_origin_lat
            home_lon = state.local_origin_lon
        else:
            home_lat, home_lon = robot_lat, robot_lon

        kml_waypoints = list(state.kml_waypoints) if state.kml_loaded else []
        kml_fence_rings = [list(ring) for ring in state.kml_fence_rings] if state.kml_loaded else []

        state.map_download_active = True
        state.map_download_status = "STARTING"
        state.map_download_error = ""
        state.map_download_path = ""
        state.map_download_started_at = time.monotonic()

    thread = threading.Thread(
        target=_satellite_download_worker,
        args=(robot_lat, robot_lon, home_lat, home_lon, kml_waypoints, kml_fence_rings),
        daemon=True,
        name="SatelliteMap",
    )
    thread.start()

    kml_note = ""
    if kml_waypoints or kml_fence_rings:
        kml_note = (
            f" + KML overlay ({len(kml_waypoints)} waypoint(s), "
            f"{len(kml_fence_rings)} fence ring(s))"
        )

    log_command(
        f"Satellite image requested: robot {robot_lat:.6f},{robot_lon:.6f} "
        f"home {home_lat:.6f},{home_lon:.6f}" + kml_note
    )
    return True


def _satellite_download_worker(robot_lat, robot_lon, home_lat, home_lon, kml_waypoints, kml_fence_rings):
    import urllib.error
    import urllib.request

    status = "ERROR"
    error_text = ""
    final_path = ""

    try:
        os.makedirs(settings.map_dir, exist_ok=True)

        rx, ry = _web_mercator(robot_lat, robot_lon)
        hx, hy = _web_mercator(home_lat, home_lon)

        # The frame must fit the KML overlay too, not just home/robot - same
        # centre+span approach, generalized from two points to however many
        # there are (with just home/robot, this reduces to exactly the old
        # calculation). A KML file can carry placemarks from more than one
        # site (e.g. a shared Google Earth project); a waypoint or fence
        # vertex tens or hundreds of km from home/robot would otherwise blow
        # the span out so far the imagery zooms out past the point of being
        # useful, so only fold in KML points within MAP_RANGE_MAX_M of the
        # home/robot area.
        ref_cx, ref_cy = (rx + hx) / 2.0, (ry + hy) / 2.0

        points_xy = [(rx, ry), (hx, hy)]
        for lat, lon, _alt, _name in kml_waypoints:
            px, py = _web_mercator(lat, lon)
            if math.hypot(px - ref_cx, py - ref_cy) <= MAP_RANGE_MAX_M:
                points_xy.append((px, py))
        for ring in kml_fence_rings:
            for lat, lon in ring:
                px, py = _web_mercator(lat, lon)
                if math.hypot(px - ref_cx, py - ref_cy) <= MAP_RANGE_MAX_M:
                    points_xy.append((px, py))

        xs = [p[0] for p in points_xy]
        ys = [p[1] for p in points_xy]

        cx = (min(xs) + max(xs)) / 2.0
        cy = (min(ys) + max(ys)) / 2.0

        span = max(max(xs) - min(xs), max(ys) - min(ys)) * MAP_IMAGE_PAD
        span = max(span, MAP_IMAGE_MIN_SPAN_M)
        half = span / 2.0

        minx, maxx = cx - half, cx + half
        miny, maxy = cy - half, cy + half

        url = (
            "https://server.arcgisonline.com/arcgis/rest/services/"
            "World_Imagery/MapServer/export?"
            f"bbox={minx:.2f},{miny:.2f},{maxx:.2f},{maxy:.2f}"
            "&bboxSR=3857&imageSR=3857"
            f"&size={MAP_IMAGE_SIZE},{MAP_IMAGE_SIZE}"
            "&format=png&transparent=false&f=image"
        )

        with state.lock:
            state.map_download_status = "DOWNLOADING"

        request = urllib.request.Request(url, headers={"User-Agent": "lazypx4"})
        with urllib.request.urlopen(request, timeout=MAP_DOWNLOAD_TIMEOUT) as response:
            data = response.read()

        if not data or len(data) < 512:
            raise RuntimeError("imagery server returned no image")

        stamp = time.strftime("%Y%m%d_%H%M%S")
        base = os.path.join(settings.map_dir, f"px4_map_{stamp}")
        final_path = base + ".png"

        def to_pixel(x, y):
            px = (x - minx) / (maxx - minx) * MAP_IMAGE_SIZE
            py = (maxy - y) / (maxy - miny) * MAP_IMAGE_SIZE
            return px, py

        home_px = to_pixel(hx, hy)
        robot_px = to_pixel(rx, ry)

        kml_waypoints_px = [
            (to_pixel(*_web_mercator(lat, lon)), lat, lon, alt, name)
            for lat, lon, alt, name in kml_waypoints
        ]
        kml_fence_rings_px = [
            [to_pixel(*_web_mercator(lat, lon)) for lat, lon in ring]
            for ring in kml_fence_rings
        ]

        # Web-Mercator distances are inflated by 1/cos(latitude); undo that
        # for real-world ground distance and scale.
        mercator_scale = max(0.05, math.cos(math.radians((robot_lat + home_lat) / 2.0)))
        distance_m = math.hypot(rx - hx, ry - hy) * mercator_scale
        ground_res = (span / MAP_IMAGE_SIZE) * mercator_scale

        annotated = False
        try:
            import io as _io

            from PIL import Image, ImageDraw

            image = Image.open(_io.BytesIO(data)).convert("RGB")
            draw = ImageDraw.Draw(image)

            def marker(point, color, label):
                px, py = point
                radius = 11
                draw.line([px - radius - 5, py, px + radius + 5, py], fill=color, width=2)
                draw.line([px, py - radius - 5, px, py + radius + 5], fill=color, width=2)
                draw.ellipse(
                    [px - radius, py - radius, px + radius, py + radius],
                    outline=color, width=3,
                )
                draw.text((px + radius + 4, py - radius - 6), label, fill=color)

            metres_per_px = ground_res
            bar_m = 25.0
            while bar_m / metres_per_px < 90.0 and bar_m < span * mercator_scale:
                bar_m *= 2.0
            bar_px = bar_m / metres_per_px
            y_bar = MAP_IMAGE_SIZE - 28
            draw.line([20, y_bar, 20 + bar_px, y_bar], fill=(255, 235, 0), width=4)
            draw.text((20, y_bar - 16), f"{bar_m:.0f} m", fill=(255, 235, 0))

            fence_color = (255, 220, 0)
            for ring_px in kml_fence_rings_px:
                if len(ring_px) >= 2:
                    draw.line(ring_px + [ring_px[0]], fill=fence_color, width=3)

            waypoint_color = (230, 60, 220)
            for i, (point, _lat, _lon, _alt, name) in enumerate(kml_waypoints_px, start=1):
                wx, wy = point
                wp_radius = 7
                label = name or f"WP{i}"
                draw.ellipse(
                    [wx - wp_radius, wy - wp_radius, wx + wp_radius, wy + wp_radius],
                    outline=waypoint_color, width=2,
                )
                draw.text((wx + wp_radius + 3, wy - wp_radius - 4), f"{i}:{label}", fill=waypoint_color)

            marker(home_px, (255, 70, 70), "HOME")
            marker(robot_px, (80, 170, 255), f"ROBOT  {distance_m:.0f} m")

            image.save(final_path)
            annotated = True
        except Exception:
            # Pillow not installed (or failed): keep the raw imagery, the
            # JSON sidecar still records where the pins belong.
            with open(final_path, "wb") as handle:
                handle.write(data)

        sidecar = {
            "captured_utc": datetime.now(timezone.utc).isoformat(),
            "source": "Esri World Imagery (server.arcgisonline.com)",
            "annotated": annotated,
            "image": os.path.basename(final_path),
            "image_size_px": [MAP_IMAGE_SIZE, MAP_IMAGE_SIZE],
            "bbox_epsg3857": [minx, miny, maxx, maxy],
            "ground_resolution_m_per_px": round(ground_res, 3),
            "home": {
                "lat": home_lat,
                "lon": home_lon,
                "pixel": [round(home_px[0], 1), round(home_px[1], 1)],
            },
            "robot": {
                "lat": robot_lat,
                "lon": robot_lon,
                "pixel": [round(robot_px[0], 1), round(robot_px[1], 1)],
                "distance_from_home_m": round(distance_m, 1),
            },
        }

        if kml_waypoints_px or kml_fence_rings_px:
            sidecar["kml"] = {
                "waypoints": [
                    {
                        "name": name,
                        "lat": lat,
                        "lon": lon,
                        "alt": alt,
                        "pixel": [round(point[0], 1), round(point[1], 1)],
                    }
                    for point, lat, lon, alt, name in kml_waypoints_px
                ],
                "fence_rings": [
                    [[round(px, 1), round(py, 1)] for px, py in ring_px]
                    for ring_px in kml_fence_rings_px
                ],
            }

        with open(base + ".json", "w", encoding="utf-8") as handle:
            json.dump(sidecar, handle, indent=2)

        status = "COMPLETE"

    except urllib.error.URLError as exc:
        error_text = f"network error: {getattr(exc, 'reason', exc)}"
    except Exception as exc:
        error_text = str(exc)

    with state.lock:
        state.map_download_active = False
        state.map_download_status = status
        state.map_download_error = error_text
        if status == "COMPLETE":
            state.map_download_path = final_path

    if status == "COMPLETE":
        note = "annotated" if final_path.endswith(".png") else ""
        if kml_waypoints or kml_fence_rings:
            note = (note + " + kml overlay").strip()
        log_info(f"Satellite image saved: {final_path} {note}".rstrip())
    else:
        log_error(f"Satellite image download failed: {error_text}")

File: lazypx4-rs/src/satellite.rs
//! Satellite-image snapshot for the mission map ([i]). Mirrors
//! `satellite.py`: one Web-Mercator export from Esri "World Imagery" (no API
//! key) covering the vehicle, home and any loaded KML overlay, annotated
//! with pins, the fence, waypoint markers and a scale bar, plus a JSON
//! sidecar recording where everything is. Background thread; needs
//! internet.

use std::f64::consts::PI;
use std::path::Path;

use image::{Rgb, RgbImage};

use crate::config::*;
use crate::geo;
use crate::state::{self, Shared, now};

fn web_mercator(lat: f64, lon: f64) -> (f64, f64) {
    let r = geo::EARTH_RADIUS_M;
    (r * lon.to_radians(), r * (PI / 4.0 + lat.clamp(-85.0, 85.0).to_radians() / 2.0).tan().ln())
}

pub fn start(shared: &Shared, map_dir: &str) -> bool {
    let job = {
        let mut st = state::lock(shared);
        if st.map_download_active {
            st.warn("A satellite-image download is already running");
            return false;
        }
        if !st.global_pos_valid {
            st.map_download_status = "ERROR".into();
            st.map_download_error = "no GPS position (need a GPS fix / GLOBAL_POSITION_INT)".into();
            st.map_download_path.clear();
            st.error("Satellite image: no GPS position available");
            return false;
        }
        let robot = (st.global_lat, st.global_lon);
        let home = if st.local_origin_set { (st.local_origin_lat, st.local_origin_lon) } else { robot };
        let kml = st.kml.clone().unwrap_or_default();
        st.map_download_active = true;
        st.map_download_status = "STARTING".into();
        st.map_download_error.clear();
        st.map_download_path.clear();
        st.map_download_started_at = now();
        let note = if kml.waypoints.is_empty() && kml.fence_rings.is_empty() {
            String::new()
        } else {
            format!(" + KML overlay ({} waypoint(s), {} fence ring(s))", kml.waypoints.len(), kml.fence_rings.len())
        };
        st.command(format!(
            "Satellite image requested: robot {:.6},{:.6} home {:.6},{:.6}{note}",
            robot.0, robot.1, home.0, home.1
        ));
        (robot, home, kml)
    };
    let (shared, map_dir) = (shared.clone(), map_dir.to_string());
    std::thread::spawn(move || {
        let (robot, home, kml) = job;
        let result = download(&shared, &map_dir, robot, home, &kml);
        let mut st = state::lock(&shared);
        st.map_download_active = false;
        match result {
            Ok(path) => {
                st.map_download_status = "COMPLETE".into();
                st.info(format!("Satellite image saved: {path}"));
                st.map_download_path = path;
            }
            Err(e) => {
                st.map_download_status = "ERROR".into();
                st.error(format!("Satellite image download failed: {e}"));
                st.map_download_error = e;
            }
        }
    });
    true
}

fn download(shared: &Shared, map_dir: &str, robot: (f64, f64), home: (f64, f64), kml: &geo::Kml) -> Result<String, String> {
    std::fs::create_dir_all(map_dir).map_err(|e| e.to_string())?;
    let (rx, ry) = web_mercator(robot.0, robot.1);
    let (hx, hy) = web_mercator(home.0, home.1);

    // Frame everything, but ignore KML geometry more than MAP_RANGE_MAX_M
    // from the robot/home area (multi-site exports would zoom out too far).
    let (ref_x, ref_y) = ((rx + hx) / 2.0, (ry + hy) / 2.0);
    let mut pts = vec![(rx, ry), (hx, hy)];
    let near = |p: (f64, f64)| (p.0 - ref_x).hypot(p.1 - ref_y) <= MAP_RANGE_MAX_M;
    pts.extend(kml.waypoints.iter().map(|w| web_mercator(w.lat, w.lon)).filter(|p| near(*p)));
    pts.extend(kml.fence_rings.iter().flatten().map(|&(la, lo)| web_mercator(la, lo)).filter(|p| near(*p)));
    let (min_x, max_x) = pts.iter().fold((f64::MAX, f64::MIN), |(a, b), p| (a.min(p.0), b.max(p.0)));
    let (min_y, max_y) = pts.iter().fold((f64::MAX, f64::MIN), |(a, b), p| (a.min(p.1), b.max(p.1)));
    let (cx, cy) = ((min_x + max_x) / 2.0, (min_y + max_y) / 2.0);
    let span = ((max_x - min_x).max(max_y - min_y) * MAP_IMAGE_PAD).max(MAP_IMAGE_MIN_SPAN_M);
    let half = span / 2.0;
    let (bx0, bx1, by0, by1) = (cx - half, cx + half, cy - half, cy + half);
    let size = MAP_IMAGE_SIZE;

    let url = format!(
        "https://server.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/export?\
         bbox={bx0:.2},{by0:.2},{bx1:.2},{by1:.2}&bboxSR=3857&imageSR=3857&size={size},{size}\
         &format=png&transparent=false&f=image"
    );
    state::lock(shared).map_download_status = "DOWNLOADING".into();
    let agent: ureq::Agent = ureq::Agent::config_builder()
        .timeout_global(Some(std::time::Duration::from_secs(MAP_DOWNLOAD_TIMEOUT_S)))
        .build()
        .into();
    let data = agent
        .get(&url)
        .header("User-Agent", "lazypx4")
        .call()
        .map_err(|e| format!("network error: {e}"))?
        .body_mut()
        .read_to_vec()
        .map_err(|e| format!("network error: {e}"))?;
    if data.len() < 512 {
        return Err("imagery server returned no image".into());
    }

    let stamp = chrono::Local::now().format("%Y%m%d_%H%M%S");
    let base = Path::new(map_dir).join(format!("px4_map_{stamp}"));
    let png_path = base.with_extension("png");

    let to_px = |(x, y): (f64, f64)| ((x - bx0) / (bx1 - bx0) * size as f64, (by1 - y) / (by1 - by0) * size as f64);
    let home_px = to_px((hx, hy));
    let robot_px = to_px((rx, ry));
    // Web-Mercator inflates distances by 1/cos(lat); undo for ground metres.
    let merc = ((robot.0 + home.0) / 2.0).to_radians().cos().max(0.05);
    let distance = (rx - hx).hypot(ry - hy) * merc;
    let ground_res = span / size as f64 * merc;

    let mut img = image::load_from_memory(&data).map_err(|e| format!("bad image: {e}"))?.to_rgb8();
    annotate(&mut img, kml, &to_px, &web_mercator, home_px, robot_px, distance, ground_res, span * merc);
    img.save(&png_path).map_err(|e| e.to_string())?;

    let r1 = |v: f64| (v * 10.0).round() / 10.0;
    let mut sidecar = serde_json::json!({
        "captured_utc": chrono::Utc::now().to_rfc3339(),
        "source": "Esri World Imagery (server.arcgisonline.com)",
        "annotated": true,
        "image": png_path.file_name().unwrap().to_string_lossy(),
        "image_size_px": [size, size],
        "bbox_epsg3857": [bx0, by0, bx1, by1],
        "ground_resolution_m_per_px": (ground_res * 1000.0).round() / 1000.0,
        "home": {"lat": home.0, "lon": home.1, "pixel": [r1(home_px.0), r1(home_px.1)]},
        "robot": {"lat": robot.0, "lon": robot.1, "pixel": [r1(robot_px.0), r1(robot_px.1)],
                  "distance_from_home_m": r1(distance)},
    });
    if !kml.waypoints.is_empty() || !kml.fence_rings.is_empty() {
        sidecar["kml"] = serde_json::json!({
            "waypoints": kml.waypoints.iter().map(|w| {
                let p = to_px(web_mercator(w.lat, w.lon));
                serde_json::json!({"name": w.name, "lat": w.lat, "lon": w.lon, "alt": w.alt, "pixel": [r1(p.0), r1(p.1)]})
            }).collect::<Vec<_>>(),
            "fence_rings": kml.fence_rings.iter().map(|r| r.iter().map(|&(la, lo)| {
                let p = to_px(web_mercator(la, lo));
                [r1(p.0), r1(p.1)]
            }).collect::<Vec<_>>()).collect::<Vec<_>>(),
        });
    }
    std::fs::write(base.with_extension("json"), serde_json::to_string_pretty(&sidecar).unwrap()).map_err(|e| e.to_string())?;
    Ok(png_path.display().to_string())
}

// ---------------------------------------------------------------------------
// Minimal raster drawing: thick lines, rings, 8x8 bitmap text.
// ---------------------------------------------------------------------------

fn plot(img: &mut RgbImage, x: i64, y: i64, c: Rgb<u8>) {
    if x >= 0 && y >= 0 && (x as u32) < img.width() && (y as u32) < img.height() {
        img.put_pixel(x as u32, y as u32, c);
    }
}

fn line(img: &mut RgbImage, (x0, y0): (f64, f64), (x1, y1): (f64, f64), c: Rgb<u8>, width: i64) {
    let steps = ((x1 - x0).abs().max((y1 - y0).abs()).ceil() as i64).max(1);
    let r = width / 2;
    for i in 0..=steps {
        let t = i as f64 / steps as f64;
        let (x, y) = ((x0 + (x1 - x0) * t).round() as i64, (y0 + (y1 - y0) * t).round() as i64);
        for dx in -r..=r {
            for dy in -r..=r {
                plot(img, x + dx, y + dy, c);
            }
        }
    }
}

fn ring(img: &mut RgbImage, (cx, cy): (f64, f64), radius: f64, c: Rgb<u8>, width: f64) {
    let n = (radius * 8.0).max(24.0) as i64;
    for i in 0..n {
        let a = i as f64 / n as f64 * 2.0 * PI;
        let mut rr = radius - width / 2.0;
        while rr <= radius + width / 2.0 {
            plot(img, (cx + rr * a.cos()).round() as i64, (cy + rr * a.sin()).round() as i64, c);
            rr += 0.5;
        }
    }
}

/// 8x8 bitmap text (ASCII), drawn with a 1 px dark outline for contrast.
fn text(img: &mut RgbImage, x: f64, y: f64, s: &str, c: Rgb<u8>) {
    use font8x8::UnicodeFonts;
    let (x0, y0) = (x.round() as i64, y.round() as i64);
    for pass in [Rgb([0, 0, 0]), c] {
        let outline = pass == Rgb([0, 0, 0]);
        for (i, ch) in s.chars().enumerate() {
            let Some(glyph) = font8x8::BASIC_FONTS.get(ch) else { continue };
            for (row, bits) in glyph.iter().enumerate() {
                for col in 0..8 {
                    if bits >> col & 1 == 1 {
                        let (px, py) = (x0 + i as i64 * 8 + col, y0 + row as i64);
                        if outline {
                            for (dx, dy) in [(-1, 0), (1, 0), (0, -1), (0, 1)] {
                                plot(img, px + dx, py + dy, pass);
                            }
                        } else {
                            plot(img, px, py, pass);
                        }
                    }
                }
            }
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn annotate(
    img: &mut RgbImage,
    kml: &geo::Kml,
    to_px: &dyn Fn((f64, f64)) -> (f64, f64),
    merc: &dyn Fn(f64, f64) -> (f64, f64),
    home_px: (f64, f64),
    robot_px: (f64, f64),
    distance: f64,
    metres_per_px: f64,
    ground_span: f64,
) {
    // Scale bar: 25 m doubled until it's at least 90 px.
    let size = img.height() as f64;
    let mut bar_m = 25.0;
    while bar_m / metres_per_px < 90.0 && bar_m < ground_span {
        bar_m *= 2.0;
    }
    let yellow = Rgb([255, 235, 0]);
    let y_bar = size - 28.0;
    line(img, (20.0, y_bar), (20.0 + bar_m / metres_per_px, y_bar), yellow, 4);
    text(img, 20.0, y_bar - 16.0, &format!("{bar_m:.0} m"), yellow);

    let fence = Rgb([255, 220, 0]);
    for r in &kml.fence_rings {
        let pts: Vec<(f64, f64)> = r.iter().map(|&(la, lo)| to_px(merc(la, lo))).collect();
        for (a, b) in pts.iter().zip(pts.iter().cycle().skip(1)).take(pts.len()) {
            line(img, *a, *b, fence, 3);
        }
    }
    let wp = Rgb([230, 60, 220]);
    for (i, w) in kml.waypoints.iter().enumerate() {
        let p = to_px(merc(w.lat, w.lon));
        ring(img, p, 7.0, wp, 2.0);
        let label = if w.name.is_empty() { format!("WP{}", i + 1) } else { w.name.clone() };
        text(img, p.0 + 10.0, p.1 - 11.0, &format!("{}:{label}", i + 1), wp);
    }
    let marker = |img: &mut RgbImage, (x, y): (f64, f64), c: Rgb<u8>, label: &str| {
        let r = 11.0;
        line(img, (x - r - 5.0, y), (x + r + 5.0, y), c, 2);
        line(img, (x, y - r - 5.0), (x, y + r + 5.0), c, 2);
        ring(img, (x, y), r, c, 3.0);
        text(img, x + r + 4.0, y - r - 6.0, label, c);
    };
    marker(img, home_px, Rgb([255, 70, 70]), "HOME");
    marker(img, robot_px, Rgb([80, 170, 255]), &format!("ROBOT  {distance:.0} m"));
}
